### PR TITLE
[Microservice] Move forward to DPM v2.0 APIs

### PR DESCRIPTION
**Pull request by chenpiaoping**
1. PM uses the new APIs of DPM
2. Remove the usage of deprecating OperationType in PM
3. Refactor update-l3-neighbors API to accept a list of gateway port IDs
4. DPM handles the configuration information of neighbors

### CODE REVIEW

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/DataPlaneClient.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
 
 @Component
 public interface DataPlaneClient {
-    List<String> createGoalStates(List<UnicastGoalState> unicastGoalStates) throws Exception;
-    List<String> createGoalStates(List<UnicastGoalState> unicastGoalStates,
-                                  MulticastGoalState multicastGoalState) throws Exception;
+    List<String> sendGoalStates(List<UnicastGoalState> unicastGoalStates) throws Exception;
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/DataPlaneClient.java:29]: This is indeed a better name.

File: web/src/main/java/com/futurewei/alcor/web/entity/route/RouterUpdateInfo.java [hunk 2]
```diff
@@ -29,8 +29,8 @@
     @JsonProperty("operation_type")
     private String operationType;
 
-    @JsonProperty("old_subnet_ids")
-    private List<String> oldSubnetIds;
+    @JsonProperty("gateway_port_ids")
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/entity/route/RouterUpdateInfo.java:32]: @kevin-zhonghao @cj-chung  FYI. Please review this change as well.

File: services/port_manager/src/main/resources/application.properties [hunk 3]
```diff
@@ -8,7 +8,7 @@ microservices.ip.service.url=http://localhost:9004/ips
 microservices.mac.service.url=http://localhost:9005/macs
 microservices.sg.service.url=http://localhost:9008
 microservices.node.service.url=http://localhost:9007/nodes
-microservices.dataplane.service.url=http://localhost:9010/v4/port/
+microservices.dataplane.service.url=http://localhost:9010/network-configuration
```
Comment (xieus) [inline services/port_manager/src/main/resources/application.properties:11]: @chenpiaoping Please also update the yaml file
https://github.com/futurewei-cloud/alcor/blob/28cfda5157c0f5e3bfaa4330c6b536217230b5d2/kubernetes/services/port_manager.yaml#L23

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java [hunk 4]
```diff
@@ -208,25 +209,51 @@ public PortWebJson getPort(String projectId, String portId) throws Exception {
...
+                }
+
+                for (PortEntity.FixedIp fixedIp: fixedIps) {
+                    String subnetId1 = fixedIp.getSubnetId();
+                    if (!StringUtils.isEmpty(subnetId1)) {
+                        gatewaySubnetIds.add(subnetId1);
+                    }
+                }
+            }
+        }
```
Comment (xieus) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java:256]: what if gatewayPortEntities ==null here? Should we add some special handling?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java [hunk 5]
```diff
@@ -108,10 +117,10 @@ private UnicastGoalState buildUnicastGoalState(NetworkConfiguration networkConfi
                     networkConfig, hostIp, portEntities, multicastGoalState));
         }
 
-        return dataPlaneClient.createGoalStates(unicastGoalStates, multicastGoalState);
+        return dataPlaneClient.sendGoalStates(unicastGoalStates, multicastGoalState);
     }
 
-    private List<String> createPortConfiguration(NetworkConfiguration networkConfig) throws Exception {
+    private List<String> processPortConfiguration(NetworkConfiguration networkConfig) throws Exception {
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:123]: Could you add brief comments to this method?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java [hunk 6]
```diff
@@ -144,20 +153,92 @@ private UnicastGoalState buildUnicastGoalState(NetworkConfiguration networkConfi
...
+        }
+
+        for (Map.Entry<String, List<NeighborInfo>> entry: hostNeighbors.entrySet()) {
+            String hostIp = entry.getKey();
+            List<NeighborInfo> hostNeighborInfos = entry.getValue();
+
+            /**
+             * At present, there are only L3 neighbors in the neighbor table,
+             * and the processing of L2 neighbors should be considered in the future.
+             */
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:209]: Can you elaborate a bit here? Neighbor table also includes L2 neighbors if my memory is right :-)

**Review by xieus — COMMENTED** [review]
@chenpiaoping Left a few minor comments. 

I linked five tracking issues to this PR based on last week's discussion. Do we cover all of them?

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java (hunk 4 above)
Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java:256]: PortController will check its legitimacy.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java (hunk 5 above)
Comment (chenpiaoping) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:123]: No problem

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java (hunk 6 above)
Comment (chenpiaoping) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:209]: In this case is for update-l3-neighbor, so there will be no L2 neighbors here.

File: services/port_manager/src/main/resources/application.properties (hunk 3 above)
Comment (chenpiaoping) [reply services/port_manager/src/main/resources/application.properties:11]: > microservices.dataplane.service.url=http://dataplanemanager-service.default.svc.cluster.local:9010/v4/port/

OK, I will do that.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java (hunk 6 above)
Comment (xieus) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:209]: aha sounds good.

File: web/src/main/java/com/futurewei/alcor/web/entity/route/RouterUpdateInfo.java (hunk 2 above)
Comment (cj-chung) [reply web/src/main/java/com/futurewei/alcor/web/entity/route/RouterUpdateInfo.java:32]: @chenpiaoping  This entity seems only used by PM's internal logic. It's a return object from updateL3Neighbors function and consumable by PM only.

**Review by cj-chung — APPROVED** [review]
LGTM

**Comment by xieus** [comment]
@chenpiaoping This PR looks good. Let us resolve the last few comments which are minor and get it merged.

**Comment by chenpiaoping** [comment]
> @chenpiaoping Left a few minor comments.
> 
> I linked five tracking issues to this PR based on last week's discussion. Do we cover all of them?

Deal with it right away.

**Comment by chenpiaoping** [comment]
> > @chenpiaoping Left a few minor comments.
> > I linked five tracking issues to this PR based on last week's discussion. Do we cover all of them?
> 
> Deal with it right away.

Resolved.

**Review by xieus — APPROVED** [review]
LGTM.

**Pull request merged**